druntime/src/core/internal/gc/blkcache.d

```diff
@@ -0,0 +1,213 @@
+
+// we expect this to be called with the lock in place
+void processGCMarks(void* data, scope IsMarkedDg isMarked) nothrow
+{
+    auto cache = cast(BlkInfo*)data;
+    // called after the mark routine to eliminate block cache data when it
+    // might be ready to sweep
+
+    debug(PRINTF) printf("processing GC Marks, %x\n", cache);
+    if (cache)
+    {
+        debug(PRINTF) foreach (i; 0 .. N_CACHE_BLOCKS)
+        {
+            printf("cache entry %d has base ptr %x\tsize %d\tflags %x\n", i, cache[i].base, cache[i].size, cache[i].attr);
+        }
+        auto cache_end = cache + N_CACHE_BLOCKS;
+        for (;cache < cache_end; ++cache)
+        {
+            if (cache.base != null && isMarked(cache.base) == IsMarked.no)
+            {
+                debug(PRINTF) printf("clearing cache entry at %x\n", cache.base);
+                cache.base = null; // clear that data.
+            }
+        }
+    }
+}
+
+unittest
+{
+    // Bugzilla 10701 - segfault in GC
+    ubyte[] result; result.length = 4096;
+    GC.free(result.ptr);
+    GC.collect();
+}
+
+/**
+  Get the cached block info of an interior pointer.  Returns null if the
+  interior pointer's block is not cached.
+
+  NOTE: The following note was not valid, but is retained for historical
+        purposes. The data cannot be cleared because the stack contains a
+        reference to the affected block (e.g. through `interior`). Therefore,
+        the element will not be collected, and the data will remain valid.
+
+  ORIGINAL: The base ptr in this struct can be cleared asynchronously by the GC,
+        so any use of the returned BlkInfo should copy it and then check the
+        base ptr of the copy before actually using it.
+  */
+BlkInfo *__getBlkInfo(void *interior) nothrow
+{
+    BlkInfo *ptr = __blkcache;
+    version (single_cache)
+    {
+        if (ptr.base && ptr.base <= interior && (interior - ptr.base) < ptr.size)
+            return ptr;
+        return null; // not in cache.
+    }
+    else version (simple_cache)
+    {
+        foreach (i; 0..N_CACHE_BLOCKS)
+        {
+            if (ptr.base && ptr.base <= interior && (interior - ptr.base) < ptr.size)
+                return ptr;
+            ptr++;
+        }
+    }
+    else
+    {
+        // try to do a smart lookup, using __nextBlkIdx as the "head"
+        auto curi = ptr + __nextBlkIdx;
+        for (auto i = curi; i >= ptr; --i)
+        {
+            if (i.base && i.base <= interior && cast(size_t)(interior - i.base) < i.size)
+                return i;
+        }
+
+        for (auto i = ptr + N_CACHE_BLOCKS - 1; i > curi; --i)
+        {
+            if (i.base && i.base <= interior && cast(size_t)(interior - i.base) < i.size)
+                return i;
+        }
+    }
+    return null; // not in cache.
+}
+
+void __insertBlkInfoCache(BlkInfo bi, BlkInfo *curpos) nothrow
+{
+    version (single_cache)
+    {
+        *__blkcache = bi;
```

Comment:
add a `return;` here. It makes it obvious that `version (single_cache)` has no further code in this function

Reply:
fixed